### ISSUE TITLE
go build error under OSX

### ISSUE THREAD

**Issue by iazkaban** [oskca/sciter#3]
# github.com/oskca/sciter
In file included from ./sciter.go:11:
In file included from include/sciter-x.h:4:
include/sciter-x-types.h:115:11: error: unknown type name 'uint16_t'
  typedef uint16_t WCHAR;
          ^
include/sciter-x-types.h:121:11: error: unknown type name 'size_t'
  typedef size_t UINT_PTR;
          ^
2 errors generated.

**Comment by oskca** [comment]
Can you post your `gcc --version`, it may relate to your gcc version. I'm compile sciter under Windows using mingw64-gcc version 5.2.0 :
```
>gcc --version
gcc (Rev3, Built by MSYS2 project) 5.2.0
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Under Linux gcc 4.8 & gcc 4.9 is fine

**Comment by iazkaban** [comment]
gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/usr/include/c++/4.2.1
Apple LLVM version 7.0.0 (clang-700.1.76)
Target: x86_64-apple-darwin15.0.0
Thread model: posix



LLVM will not work well?

**Comment by oskca** [comment]
As I have no Mac machine, the OSX version is not tested yet. So I don't konw whether it would work or not. :)

Does the sciter-sdk examples work under OSX? If it works I think there's some improvements need to be done for OSX

2015-10-30 

datago 



发件人：隋龙飞 <notifications@github.com>
发送时间：2015-10-30 14:38
主题：Re: [sciter] go build error (#3)
收件人："oskca/sciter"<sciter@noreply.github.com>
抄送："oskca"<datago@yeah.net>

gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/usr/include/c++/4.2.1
Apple LLVM version 7.0.0 (clang-700.1.76)
Target: x86_64-apple-darwin15.0.0
Thread model: posix
LLVM will not work well?

**Comment by ryuken** [comment]
I got the same problem on OSX.
With sciter-sdk examples do you mean the demo.osx folder?

**Comment by oskca** [comment]
Yes! Can you build the sdk example using LLVM under OSX?

**Comment by ryuken** [comment]
I just tested the samples: [layered, respack, sciter]
They all compiled and ran (except for respack because it's a command line tool i think)

**Comment by oskca** [comment]
Well, then it seems there's something wrong with the header files provided by this project.

Possible solutions:

1. `add #include <stddef.h> in include/sciter-x-types.h`

2. if that won't work, the problem may relates to `ORDER` of the included headers, you can have a look at [this link](http://cncc.bingj.com/cache.aspx?q=llvm+gcc+unknown+type+name+%27uint16_t%27&d=4877980542501059&mkt=zh-CN&setlang=zh-CN&w=ak4uQBY23MmqP4OhIYhCGeztVgEBvd-q) for a quick hint.

3. and finally if the previous two both failed, try adding 
    ```
   typedef long unsigned int size_t;
   typedef unsigned short  uint16_t;
    ```
   directly to  include/sciter-x-types.h

if you solve this problem can you create a pull request :smile:

**Comment by ryuken** [comment]
Thanks for your quick response :D
The first option didn't work.

I couldn't click the link in the 2nd option, but found a comparable link i think.
Someone had the same problem in another project and he included stdio.h + stdint.h
I included these in sciter-x.h and got the following problem:

../../sciter.go:882:8: error: field has incomplete type 'VOID' (aka 'void')
../../sciter.go:1155:8: error: field has incomplete type 'VOID' (aka 'void')

I also get the same problem when using option 3.

**Comment by oskca** [comment]
The link in 2nd option, is about  unknown type name 'size_t', here is it's content:

 “error: unknown type name 'size_t'” solved while compiling withllvm-gcc
2013-08-22 10:58 1974人阅读 评论(1) 收藏 举报
I went into erros while compiling the very simple code below with llvm-gcc:


[cpp] view plaincopy#include <stdio.h>  
int main(){  
    printf("nani\n");  
    return 0;  
}  


I got error like: “error: unknown type name 'size_t'” 


It seems that llvm-gcc cannot find the header file which defines size_t(actually the file is <stddef.h>).


I thought it was easy to settle down and added C_INCLUDE_PATH=/usr/include/:/usr/include/linux:/usr/include/i386-linux-gnu to the environment profile.
However, the errors occur still. I was confused by this because I have told llvm-gcc where to locate the header files it needs explicitly by the varible defined in environment profile.
Then I searched for <stddef.h> and found there are several versions of this file and each of them are located in different directories- one of them in /usr/include/liunx and one in another. I opened the versions of the file one by one to find the content of the files are different !!! Actually the one in /usr/include/linux DOES NOT define size_t at all !!!!


Now things get clear. By putting /usr/inlcude/linux in front of the path where the effective <stddef.h> (which defines size_t) is located, I made the compiler use the first one it encountered, that is,  /usr/include/linux/stddef.h. But there is nothing in this <stddef.h>, at least not the things the compiler needs. 


So, the very correct way to solve this problem is adding header files searching paths in the RIGHT ORDER to the environment profile. Here, I did not add /usr/include/linux at all. 


Problem solved.


And for gcc users who encounter the same all similar problems, this solution also works, I think. 

Would this help?

**Comment by ryuken** [comment]
It resolved that problem but gives the following problem (same as other options):
../../sciter.go:882:8: error: field has incomplete type 'VOID' (aka 'void')
../../sciter.go:1155:8: error: field has incomplete type 'VOID' (aka 'void')

I don't speak C/C++ very good so I don't really understand this error

**Comment by ryuken** [comment]
I found this. Is it helpful? 

http://stackoverflow.com/questions/9742135/g-error-anonymous-has-incomplete-type

**Comment by oskca** [comment]
This's wired. The errors are the C compiler errors but happend in Go files. The link you provided indicates something relates to C++ compiler but this cgo project has no C++ code at all.

I can't find the proper places where these errors are located in sciter.go. The line near line 882 has code bellow:
```Go
func (e *Element) Update(forceRender bool) error {
	// args
	var cforceRender C.BOOL = C.FALSE
	if forceRender {
		cforceRender = C.TRUE
	}
	// cgo call
	return wrapDomResult(C.SciterUpdateElement(e.handle, cforceRender), "SciterUpdateElement")
}
```
whic involves no `VOID` things and so for line 1155. And the cgo generated files has no such things either.

As I have no OSX machine I can't figure out what this error exactly means. :sweat:

**Comment by therealbill** [comment]
I can get it to compile fine w/some changes. The above includes, but also converting some VOID definitions to LPVOID. There are implementations for _windows and _linux, however. I can  make a PR if desired to clarify those changes.  So, yay.

Unfortunately there doesn't seem to be the full window definition in sciter/window/window.go.  Thus examples/simple doesn't actually build because of the lack of window.New. Working on it but as I am sure you know, usable Sciter documentation is severely lacking or difficult to find.

**Comment by ryuken** [comment]
It would certainly help me if you make a PR :P

**Comment by ryuken** [comment]
Just to be sure. You compiled it on OSX right? Not on Windows?

**Comment by oskca** [comment]
Well, OSX is not supported yet (I've updated the README), but PRs would be of great help :)

**Comment by therealbill** [comment]
Yeah, I compiled it on OS X. I'll submit a PR for those changes later today. Been digging around to see what to do to get a Window to show (ie. the "example/simple" demo) from Sciter/C on OS X.

**Comment by oskca** [comment]
It seems that `C.SciterCreateWindow` works under OSX. Though the function arguments vary for each platform `nil` seems just ok :smile:  from the examples in  `demo.osx` :

     HWINDOW hw = SciterCreateWindow(SW_MAIN | SW_ALPHA | SW_TOOL, &frame, 0,0,0);

Be careful with the `sciter.WindowCreationFlag`, I used to create 1px sized window under Linux which was really hard to find :smile:

**Comment by therealbill** [comment]
"Be careful with the sciter.WindowCreationFlag, I used to create 1px sized window under Linux which was really hard to find"

I bet!

yeah the create window works fine but I haven't found the equivalent calls for implementing .SetTitle() .Show() and .Run(). Yet.

**Comment by kuznetsovin** [comment]
I created a pull request #31, which fixes the problem with .SetTitle() .Show() and .Run() functions, I tested on my OS X 11.11 all examples working fine.

**Comment by oskca** [comment]
fixed by #31
